### ISSUE TITLE
Add unicode input method for turnstile characters?

### ISSUE THREAD

**Issue by emacsomancer** [greghendershott/racket-mode#393]
The [turnstile language](https://docs.racket-lang.org/turnstile/The_Turnstile_Reference.html#%28part._.Typing_.Unicode_.Characters%29) uses six additional special unicode characters:

- \vdash → ⊢
- \gg → ≫
- \rightarrow → →
- \Rightarrow → ⇒
- \Leftarrow → ⇐
- \succ → ≻

`racket-unicode-input-method.el` looks easy enough to extend for these, but would a pull request be worthwhile in this case? Or should  `racket-mode` be kept restricted to the current set of special characters?

**Comment by stchang** [comment]
FWIW, I just manually modify `racket-unicode-input-method.el`, as you mentioned. Though I agree a better solution would be nice.

@greghendershoot, do you know if this situation is a use case for a minor mode? (I know nothing about emacs modes)

**Comment by capfredf** [comment]
here is how I add one turnstile characters to racket-unicode-input-method:

```elisp
;; (require 'racket-mode) or put the code below in the `:config` part if you use `use-package`
(with-temp-buffer
    (set-input-method "racket-unicode")
    (let ((quail-current-package (assoc "racket-unicode" quail-package-alist)))
      (quail-define-rules ((append . t))
                          ("vdash" ["⊢"])
                          ("^o" ["ᵒ"]))))
```

**Comment by emacsomancer** [comment]
@capfredf That also suggests a possible method of extending base `racket-mode` to allow these to be part of the `racket-mode` package itself, but leave them off by default (i.e. have a package variable like `use-turnstile-chars` which defaults to `'nil`, but when non-nil appends the turnstile special chars to the base `racket-unicode` set).

**Comment by greghendershott** [comment]
- Back when I cargo-culted this from Haskell Mode, either Turnstile didn't yet exist or I wasn't aware of it. Otherwise I think I would have supplied these, too. Thanks for suggesting this!

- I'd be happy to add them, now. (Or soon-ish, after I finish tweaking things for #392 which is turning out to be a "pull the thread and the sweater unravels" thing.)

- Being Emacs, there are probably 42 ways to accomplish this. Newer Emacs have a prettify-symbols feature, for instance. I'm not immediately sure all the pros and cons, and whether Racket Mode even "needs" to supply this, anymore. (In any case I'd keep supporting it. Just not sure whether to recommend vs. deprecate it. I don't know yet.) 

- Thanks @capfredf for sharing how to add more to the defaults. Can people remove defaults that way, too? If so, then @emacsomancer maybe the simplest thing would be to document that approach [here](https://www.racket-mode.com/#Unicode-input-method) and [here](https://www.racket-mode.com/#racket_002dunicode_002dinput_002dmethod_002denable).

- Longer term, if/as practical, Racket Mode should probably evolve to being more of a "Racket is a programming language programming lanague mode" -- for example using font-lock and indent supplied by lang-info. Perhaps these kinds of shortcuts should be another category of stuff that a lang may supply?? I don't know for sure.

TL;DR: I'll add these Turnstile items to the defaults, and, document how to add to (and remove from?) the defaults.

**Comment by greghendershott** [comment]
> Being Emacs, there are probably 42 ways to accomplish this. Newer Emacs have a prettify-symbols feature, for instance. I'm not immediately sure all the pros and cons, and whether Racket Mode even "needs" to supply this, anymore. (In any case I'd keep supporting it. Just not sure whether to recommend vs. deprecate it. I don't know yet.)

Actually I glossed over an important difference:

- an input-method will *replace* the text -- for instance "gg" with "≫".
- a prettify symbols feature will keep the original text but use an Emacs overlay to _display_ as the other text

So apples and oranges, each has its own pros and cons.

---

Anyway, I finished up #392 and can work on this next:

>  I'll add these Turnstile items to the defaults, and, document how to add to (and remove from?) the defaults.

**Comment by emacsomancer** [comment]
>  Can people remove defaults that way, too? 

Looking at quail, there's an obvious way to append, but not to remove.

> Actually I glossed over an important difference:

>    an input-method will replace the text -- for instance "gg" with "≫".
>    a prettify symbols feature will keep the original text but use an Emacs overlay to display as the other text

> So apples and oranges, each has its own pros and cons.

I think `turnstile` would require `(require turnstile/no-unicode)` for the prettify symbols route.

**Comment by greghendershott** [comment]
@emacsomancer Good points. Thanks.

**Comment by greghendershott** [comment]
I just pushed commit 51207f3 to a topic branch.

- Do the original strings make sense? (I didn't use a leading `\` to be consistent with the existing ones -- is that OK?).

- Is this tip a reasonable hint?  (Note this would appear in the doc string, Info, and HTML at https://www.racket-mode.com).

**Comment by emacsomancer** [comment]
@greghendershott 

> Do the original strings make sense? (I didn't use a leading \ to be consistent with the existing ones -- is that OK?).

I think so. That's how I did it in my ad hoc implementation.

> Is the tip a reasonable hint? (Note this would appear in the doc string, Info, and HTML at https://www.racket-mode.com).

As per https://github.com/greghendershott/racket-mode/issues/393#issuecomment-512508902, at least with my set-up, I had to add `(racket-unicode-input-method-enable)` within the temp-buffer to have it work.

**Comment by greghendershott** [comment]
Unless anyone else has any feedback, I'll go ahead and merge commit 51a340f to master this evening.

**Comment by stchang** [comment]
> Unless anyone else has any feedback

These are not Turnstile bound ids, but other symbols (not already included by `racket-mode`) commonly used in the language of type rules may include:

- × `times`
- ⇑ `Uparrow`
- ⇓ `Downarrow`
- ∨ `vee`
- ∧ `wedge`
- ⊏ `sqsubset`
- ⊐ `sqsupset`
- ⊑ `sqsubseteq`
- ⊒ `sqsupseteq`
- ⊓ `sqcap`
- ⊔ `sqcup`
- ⊧ `models`

While I would love if these were included as well, I realize it's easy to get carried away, so I would be fine if it's decided to leave them out.

**Comment by greghendershott** [comment]
@stchang Good idea; I added those in commit 83df886.

**Comment by greghendershott** [comment]
Before I merge, one (last?) question: Should these have trailing spaces? IIUC that means for example you would need to type "gg " with the benefit that it wouldn't match on "eggo".

I didn't think about that too carefully when I cargo culted this from Haskell Mode years ago. Mostly those had trailing spaces, except for things with "natural delimiters" like "|A|". And yet, there were a couple inconsistencies like "forall" and "exists", no trailing space. To which I added more, like the Redex ones. And now we're adding even more.

It seems like all except things like "|A|" ought to have trailing spaces?

I _think_ that makes sense but I don't use this feature for real myself, just in testing. So I'd be more comfortable hearing from those of you who do use it, including @stchang  @emacsomancer @capfredf.

**Comment by stchang** [comment]
Just to clarify, "trailing spaces" means that it won't convert to the symbol until you type the letter sequence *and* press a whitespace (ie, the current behavior for the greek letters)?

If so, I think that's fine.

**Comment by emacsomancer** [comment]
I did think about this, specifically with respect to "gg", which seems like it could match in unwanted places.

I haven't used racket-mode's unicode replacing feature for long enough to have a firm opinion. It makes sense though, I think.

**Comment by greghendershott** [comment]
@stchang Yes, exactly.

@emacsomancer Thanks. Especially for existing ones like "forall", I didn't want to change and create an annoyance I didn't realize.

I'll make those changes, and unless @capfredf weighs in with some new consideration, plan to merge later today or tomorrow-ish.
